### ISSUE TITLE
Windows_Build failing again

### ISSUE THREAD

**Issue by bscottm** [open-simh/simh#265]
According to open-simh's actions, libpng16 is the culprit again.

**Comment by pkoning2** [comment]
It's still broken.  Can anything be done about this short of removing Windows from the CI list?

**Comment by markpizz** [comment]
I sent the following message about this to Timothe Litt on 7/5/2023 and haven't heard anything back from him:
```
Tim,

Your windows_build.yml script depends on the legacy windows-build repo from https://github.com/simh/windows-build.
You've got it building with the VS2022 compiler that is still in its development phase and has updates from Microsoft sometimes every 6 months, sometimes monthly.  When an updated version of VS2022 comes out, some (but strangely not all) static libraries compiled in Release mode are incompatible with prior libraries.  The VS2022 compiler and, in general, the GitHub actions "latest" images are a constantly moving target since they update the tool chain at arbitrary times.  When this happens, your existing CI build fails since the windows-build libraries don't precisely track the moving targets in the build platform.

Since VS2008 (some 15 years back), nothing has changed in the various Visual Studio versions that have enhanced anything about the capabilities of the running simulators, so nothing drives using the absolute latest Visual Studio.

I would suggest that you modify this CI script so that it uses a VS2019 version which is stable now that VS2022 is the one being changed on the fly.  That will therefore be stable and avoid the random CI failures that are happening now and will most likely continue work just fine for a good number of years.

Additionally, when you set this stuff up, you added yet another maintenance point for folks adding new modules to any simulator.  Before, we had:
- makefile
- Visual Studio Projects\simh.sln and .vcproj files
- descript.mms

Now there is your extra stuff supporting VS2022 in the .vcxproj files that needs to be maintained by hand.

Microsoft's Visual Studio compiler suites have been 100% upward compatible with being able to migrate from the older version configuration info into each of the latest Visual Studio versions.  As a result, I always avoided having extra .vcxproj files to maintain since merely opening the older simh.sln file (or a copy of that) would automatically convert it and all of the contained .vcproj references, to .vcxproj files and that would just work on the invoking Visual Studio version.  Working from this concept, there exists a script already in the repo that will both convert and build on essentially newer Visual Studio versions, and should be directly callable from a slightly modified CI script and you can remove all the .vcxproj files and the simh.ci.sln file.

The necessary build step would look something like this:
          build_vstudio.bat

This will work for VS2019, VS2022 and other VS versions.  It looks first for an installed VS2008 version or the most recent VS version installed on the host system.  If it is run on a system with VS2022 installed, you'll get the same failure that's happening now.

You DO NOT need to do anything about providing windows-build library stuff.  The build scripts will download windows-build automatically.
Executing the build_vstudio.bat will overwrite "Visual Studio Projects\simh.sln", but since this is going on in the CI context, local changes disappear at the end of the CI run anyway.
All the .vcxproj files will need to be removed from the repo before the run since the conversion process will create prompts before overwriting any local .vcxproj files.  The existing .vcxproj file seem to have some nuts and bolts related to building on ARM, but that can't possibly work since there is no Windows-Build support for ARM at this time.

I notice that building with this method also builds the recently added NT100 simulator which adds fuel to my above argument about excessive places to update build info since that new simulator was never added to the .vcxproj stuff.

It seems that the open-simh windows builds sometimes randomly hangs during the testing of at least one simulator.  The tests are hanging can be skipped for all simulators if prior to executing the above build_vstudio.bat file, you create a file called "Visual Studio Projects\ Post-Build-Event.Skip" with anything in it.

Good Luck.

- Mark
```

**Comment by pkoning2** [comment]
So MS doesn't even support binary compatibility across service packs for a given compiler?  What a pack of morons.

**Comment by pkoning2** [comment]
What you say makes perfect sense, but since I avoid WIndows at all cost I have no idea how to do that.  I'm looking for someone who knows.  Anyone?

**Comment by pkoning2** [comment]
So if the cmake flow works around MS design errors, while the other one does not, it would make sense to stop using the other one.  It doesn't appear to add anything critical to the CI flow.

**Comment by markpizz** [comment]
> @pkoning2 So if the cmake flow works around MS design errors, while the other one does not, it would make sense to stop using the other one. It doesn't appear to add anything critical to the CI flow.

Absolutely true, I'm unaware of any discussion at all about Timothe Litt's adding of the CI work he put together and general build VS2022.  If the CI stuff goes, then the "Visual Studio Projects\*.vcxproj" and "Visual Studio Projects\Simh.ci.sln" files should go as well.  Otherwise they are another maintenance place for build instructions.  These extra files were never needed since they're generated from the existing "Visual Studio Projects\*.vcproj" and "Visual Studio Projects\Simh.sln" files when the Simh.sln file is opened by any newer version of Visual Studio. 

> @bscottm Does MS maintain their product and provide bug fixes? Yes -- saying that VS 2022 is "in development" is counterfactual to a released product.

Well, their "fixes" break pre-existing code every time they deliver fixes.  This would potentially be expected by something in development, but certainly not anything that is formally released.  This breakage problem has been consistently how they've done things since at least VS2017 I never noticed before then, but I rarely had anyone wanting any newer VS version.

Talking about different Visual Studio releases, I'd like you to elaborate on what you see as the advantages of VS2022 over any prior VS version going back to VS2008.  As far as functionality needed by the simh project, all prior VS releases have consistently produced fully functional simulator code.  Clearly, very old versions are potentially hard to support forever, but why is VS2022 better for simh than VS2019?

**Comment by bscottm** [comment]
> > @pkoning2 So if the cmake flow works around MS design errors, while the other one does not, it would make sense to stop using the other one. It doesn't appear to add anything critical to the CI flow.

CMake doesn't work around MS design errors. CMake builds from source when necessary. "Necessary" means "on Windows native and XP-compatible builds."

I intentionally chose to build dependency libraries from source because:

* it ensures a consistent result because you know what you're getting,
* it gracefully deals with environment changes,
* doesn't require curating and maintaining yet another series of artifacts, and 
* it's easier to upgrade to new library versions (just change the version number, fetch the tar.gz or zip archive.)

Were there a need to do so, I'd have done the same thing on Linux and Mac. Fortunately, Debian and Mac have trustworthy package managers, so there's no need to build dependency libraries from source.

> Talking about different Visual Studio releases, I'd like you to elaborate on what you see as the advantages of VS2022 over any prior VS version going back to VS2008. As far as functionality needed by the simh project, all prior VS releases have consistently produced fully functional simulator code. Clearly, very old versions are potentially hard to support forever, but why is VS2022 better for simh than VS2019?

_"Talking about different GCC releases, I'd like you to elaborate on what you see as the advantages of GCC 12 over any prior GCC version going back to GCC 4. As far as functionality needed by the simh project, all prior GCC releases have consistently produced fully functional simulator code. Clearly, very old versions are potentially hard to support forever, but why is GCC 12 better for simh than GCC 4?"_

You get whatever GCC version that's distributed on the platform. Same goes for Visual Studio on GH -- VS 2022 Enterprise is what's installed on the GH Windows runner. Installing a previous version is righteous pain. It's less painful on Appveyor because they have CI/CD images with older VS environments (although Appveyor is a righteous pain in its own way.)

There's no requirement for you personally to use VS 2022 in your own environment. It's not clear to me that there's any singular advantage to using a 15 year old compiler and development platform just for some level of bugward and forward compatibility in the VS project and solution files.

Really need VS 2008 project and solution files? CMake can generate those, which is documented.

**Comment by markpizz** [comment]
> > Talking about different Visual Studio releases, I'd like you to elaborate on what you see as the advantages of VS2022 over any prior VS version going back to VS2008. As far as functionality needed by the simh project, all prior VS releases have consistently produced fully functional simulator code. Clearly, very old versions are potentially hard to support forever, but why is VS2022 better for simh than VS2019?
>
>"Talking about different GCC releases, I'd like you to elaborate on what you see as the advantages of GCC 12 over any prior GCC version going back to GCC 4. As far as functionality needed by the simh project, all prior GCC releases have consistently produced fully functional simulator code. Clearly, very old versions are potentially hard to support forever, but why is GCC 12 better for simh than GCC 4?"
> 
> You get whatever GCC version that's distributed on the platform. Same goes for Visual Studio on GH -- VS 2022 Enterprise is what's installed on the GH Windows runner. Installing a previous version is righteous pain. It's less painful on Appveyor because they have CI/CD images with older VS environments (although Appveyor is a righteous pain in its own way.)
>
> There's no requirement for you personally to use VS 2022 in your own environment. It's not clear to me that there's any singular advantage to using a 15 year old compiler and development platform just for some level of bugward and forward compatibility in the VS project and solution files.

Notice that I asked absolutely nothing about GCC.

In any case, various CI images provide different VS versions.  I’m not suggesting that any version of VS get installed, but that the CI Image with VS2019 be used since this issue is specifically about that CI activity which was initially setup @tlhackque.

> Really need VS 2008 project and solution files? CMake can generate those, which is documented.

AGAIN: The CI build setup by @tlhackque explicitly uses VS projects that were created by opening the existing  VS2008 solution file.  I'm saying that ANY version of  VS2010-VS2022 will produce these project files any time the VS2008 ones are opened, and therefore these are "derived" or "generated" project files and shouldn't exist in the repo since they add no value, but are yet another maintenance point for folks who add a new module to a simulator.  We don't store .obj or .o files in the repo, right?

This discussion IS NOT about your cmake activities.

**Comment by bscottm** [comment]
@pkoning2: Respectfully request permission to drop the `windows_build.yml` workflow, since the CMake workflow builds and packages Windows executables. Avoids this whole complicated issue in its entirety.

**Comment by pkoning2** [comment]
Sounds good to me.  Can you generate a PR that does this?

**Comment by bscottm** [comment]
@pkoning2: PR #284 submitted.

**Comment by pkoning2** [comment]
Closed by PR #284 merge.
